### ISSUE TITLE
Persistent issues with graceful server shutdown on multiple platforms

### ISSUE THREAD

**Issue by MihaelIsaev** [vapor/vapor#2502]
It seems that in Vapor 4.30.0 I can't stop the app by pressing Ctrl+C in console. 
It just shows ^C and doing nothing. After several attempts I decided to downgrade to 4.29.1 and Ctrl+C works fine on it. 
To make sure my theory I updated back to latest 4.30.0 and Ctrl+C is broken again here. 

### Steps to reproduce

1. define `.exact("4.30.0")` for vapor package in `Package.swift`
2. run `swift package update && swift run`
3. try to terminate the app by pressing `Ctrl+C` and see that it just prints `^C` on the screen
4. then define `.exact("4.29.1")` for vapor package in `Package.swift`
5. run `swift package update && swift run`
6. try to terminate the app by pressing `Ctrl+C` and it terminates the app

### Expected behavior

`Ctrl+C` should terminate the app

### Actual behavior

It just prints `^C` on the screen and doesn't terminate the app

### Environment
* Vapor Framework version: 4.30.0
* Vapor Toolbox version: n/a
* OS version: Ubuntu 20.04
* Swift version: swift-5.2.5-RELEASE

**Comment by fouadhatem** [comment]
@tanner0101 I was able to recreate this as well on all supported Ubuntu versions `16.04`, `18.04` and `20.04` with Swift `5.25` and Swift `5.3`. Furthermore, `CTRL-C` in `4.31.0` displays the prompt without actually terminating the process. Running `ps` after `CTRL-C` will show you the process continuing to run in the background. This can also be confirmed by attempting to run the server again on the same port after `CTRL-C`, which will result in an `Address already in use (errno: 98)` error. I don't think this has to do with an earlier similar error which was triggered by open web sockets. Graceful shutdown seems to have been broken somewhere since that fix and for a different reason this time

Whatever is triggering this may also explain an unexplained error that I and a handful of other people also seem to be having on the Mac. Previously, I randomly received a `-1` exit status code instead of the typical `9` exit status code I received from Xcode when terminating the project. Whenever that happens, I have to manually terminate the `Run` process in order to be able to run the server again on the same IP:port combination with the `-b` option. Now, it happens each and every time I run the project.

**Comment by Maxim-Inv** [comment]
Here is lines from supervisord.log when I tried to stop Vapor 4.29.4 on Ubuntu 18.04:
`2020-10-05 09:57:02,268 INFO waiting for vapor to stop
2020-10-05 09:57:02,291 INFO stopped: vapor (exit status 0)`

and Vapor 4.30.0:
`2020-10-05 09:56:23,286 INFO waiting for vapor to stop
2020-10-05 09:56:25,286 INFO waiting for vapor to stop
2020-10-05 09:56:27,286 INFO waiting for vapor to stop
2020-10-05 09:56:29,286 INFO waiting for vapor to stop
2020-10-05 09:56:31,286 INFO waiting for vapor to stop
2020-10-05 09:56:33,286 WARN killing 'vapor' (25243) with SIGKILL
2020-10-05 09:56:33,286 INFO waiting for vapor to stop
2020-10-05 09:56:33,294 INFO stopped: vapor (terminated by SIGKILL)`

It seems that vapor doesn't accept/listen the SIGTERM signal.

**Comment by fouadhatem** [comment]
@MihaelIsaev I suggest you rename the topic of this issue to: Persistent issues with graceful server shutdown on multiple platforms.

**Comment by fouadhatem** [comment]
This is baffling. I just tried the following:

- I built a skeleton project on a laptop running a desktop version of Ubuntu 20.04 with 4GB of memory and graceful shutdown worked fine each time.

- I ran the same skeleton project on a 2019 MacBook Pro with 16GB of memory running Catalina and the project terminates just fine on Xcode with exit code 9 and not with a -1 and an RPC server failure as I consistently get on my own MacBook (2013 MacBook Air with 4GB of memory running Catalina).

- When I ran the steps above to reproduce the error that @MihaelIsaev filed the issue for, I was able to reproduce it each time with the following setup: I am running VMWare on my MacBook Air (4GB of RAM). The tests were conducted on Ubuntu Server VMs (varying flavors) with 1 processor core and 1GB of RAM for each VM. The running process could not be terminated with `CTRL-C` or with `kill <PID>`. It could only be killed with `kill -9 <PID>`. Attempting to build and run again on those VMs result in the same scenario and needed another `kill -9` to end the process.

- Today, I picked one of these VMs, added a processor core and 1GB of RAM to its existing configuration and ran the same skeleton project and it initially failed to gracefully shut down. However, after only one `kill -9` and a 2nd attempt to run, it terminated fine and continued to do so every after I have rebooted the device. Here's the kicker, shrinking the VMs config back down to 1 processor core and 1GB of RAM seemed to cause the same issue to recur again. 

This might be a problem that is specific to machine, or it might indicate a pattern where running the project on machines with specific configurations results in this kind of behavior. The variations I can think of is limited or processing capability, with my hunch leaning more towards memory issues.

This might explain why @tanner0101 and @0xTim were unable to reproduce these errors, as they are probably using far more capable Macs with generous helpings of memory and abundant processing power. Could be a `Swift-NIO` thing. Also, reporting of similar errors were sporadic and far between, which leads me to believe that this may be the case, unless I'm missing something.

@Maxim-Inv and @MihaelIsaev can you please share the configurations of your computers/VMs that you were using to reproduce these errors?

**Comment by MihaelIsaev** [comment]
@fouadhatem Thank you for the investigation!

I'm not able to reproduce it on macOS, Ctrl+C works here. My machine is MacBook Pro 2017 15" 16Gb RAM.

My Ubuntu server is 1Gb cheapest machine on DigitalOcean with Ubuntu 20.04. When I run my app there are 800Mb of free RAM.

**Comment by fouadhatem** [comment]
So the issue appears on your DigitalOcean VM running on 1GB of RAM? What kind of processing power does it have and how many cores does it give you?

**Comment by fouadhatem** [comment]
Alright I'm now pretty sure that this is a resource allocation issue, most likely some kind of memory management problem. I decided to test this in a more standardized approach, using EC2 instances on AWS. Below are the details of the test:

### Methodology

I created 4 instances with the following configurations:

1. Ubuntu 20.04 Server instance with 1 processor and 1GB of RAM.

1. Ubuntu 20.04 Server instance with 1 processor and 2GB of RAM.

1. Ubuntu 20.04 Server instance with 2 processors and 4GB of RAM.

1. Ubuntu 20.04 Server instance with 2 processors and 1GB of RAM.

All instances had 30GB of SSD storage. After updating each instance with `apt update` and `apt upgrade`, installing `Swift 5.3` and its dependencies and the Vapor toolbox, I executed the following sequence of commands on each instance:

1. `vapor new -n Mark1`

1. `cd Mark1`

1. `vapor build`

1. `vapor run`

1. `CTRL-C`

1. `ps` // to see if the `Run` process is still there.

1. `vapor run` // to see if we can run the server again after `CTRL-C`. Do not proceed if successful.

1. `kill <PID>` // to attempt to kill the `Run` process, if it's still running.

1. `ps` // to see if we managed to kill it. Proceed if it still exists.

1. `kill -9 <PID>` // to put a bullet between its eyes.

1. `ps` // to make sure it's really dead this time.

1. `vapor run`

1. `CTRL-C`

1. `ps`

### Results

The results are below for each instance, showing you the screen from Step 4 onwards. Some steps were repeated on some instances for demonstration purposes:

### One processor with 1GB of memory - failed to terminate

![1P1G](https://user-images.githubusercontent.com/24614177/95406105-4967a780-0922-11eb-8f69-09c7eb0bc42d.png)

### One processor with 2GB of memory - failed to terminate

![1P2G](https://user-images.githubusercontent.com/24614177/95406126-58e6f080-0922-11eb-83f8-46ea8f702a5b.png)

### Two processors with 4GB of memory - successful

![2P4G](https://user-images.githubusercontent.com/24614177/95406154-67350c80-0922-11eb-8986-a201b885acb9.png)

### Two processors with 1GB of memory - fails to build project repeatedly

It kept failing at `PasswordHasher.swift`, even after a reboot.

![2P1G](https://user-images.githubusercontent.com/24614177/95407919-ffcd8b80-0926-11eb-931a-25c46f43180b.png)

### Conclusion

I am able to reasonably conclude that this seemingly minor problem has uncovered some kind of resource management issue (most likely memory) with either Vapor or Swift-NIO, which could also be related to #2493 .

Hope this helps!

**Comment by MihaelIsaev** [comment]
@fouadhatem Memory issue sounds weird since Vapor app eats only 12Mb of RAM 🤷‍♂️

I tested that theory with my apps on DigitalOcean by increasing resources of droplet from 1 to 5

1) Shared CPU 1 vCPU 1 GB RAM $5/mo (my regular droplet)
2) Shared CPU 1 vCPU 2 GB RAM $10/mo
3) Shared CPU 2 vCPUs 2 GB RAM $15/mo
4) Shared CPU 2 vCPUs 4 GB RAM $20/mo
5) Shared CPU 4 vCPUs 8 GB RAM $40/mo

Results:
1,2,3) Ctrl+C and kill failed, kill -9 works
4,5) Ctrl+C and kill works

So yeah, I can confirm that on VPS with more resources it works for some reason 🧐

**Comment by Maxim-Inv** [comment]
@fouadhatem I encountered this issue on AWS t3.small server (2 vCPU, 2 GB RAM). And it works fine on the much more might server with 64 GB RAM.

**Comment by fouadhatem** [comment]
@MihaelIsaev and @Maxim-Inv thanks for confirming the tests on Linux. The same issue seems to happen with me on my 2013 MacBook Air, which has 4GB of RAM. When I try it on a more capable Mac, the issue goes away. Does anyone have an older Mac lying around that we can test this with?

**Comment by jhoughjr** [comment]
Just wanted to chime in that my iMac has a quad core i7 4.0 with 32 GB of ram. I have still randomly 
encountered this issue. I also encountered it on my droplets, whether 1 vCPU or 2, 1 GB RAM or 4 GB RAM.
In my tests it seemed related to if the server received requests or not, but I was not able to confirm that.
I don’t think memory allocation is the issue.

**Comment by fouadhatem** [comment]
@jhoughjr how long ago was this? There indeed was a previous issue with the server failing to shutdown when it had received requests (see #2451). This was resolved in 4.28.0. Could that be what you were referring to? That issue was not related to memory or resource capacities as this one seems to be.

**Comment by fouadhatem** [comment]
I think I need to bring in some heavy guns over from Discord for this...

**Comment by fouadhatem** [comment]
Okay so long story short, this seems to be a symptom of a larger and deeper problem that has to do partly with Vapor and partly with Swift-NIO. It is currently being evaluated by someone at Swift-NIO. Expect a PR to address this in the coming days (or weeks, depending on how deep it is).

**Comment by jhoughjr** [comment]
I dont think I have seen it simce then but ive not been working on server much.
I can try this weekend to see if it still happens.

**Comment by mikecsh** [comment]
Just to add a datapoint, I am consistently seeing this both in Xcode 12.0.1 and in Ubuntu 18.04 under Docker (Engine v19.03.13) on a 2017 1.4 GHz Dual-Core Intel Core i7 MacBook 12" with 8 GB RAM running macOS 10.15.4.

**Comment by fouadhatem** [comment]
@mikecsh yeah the issue seems to have been tracked down, at least partially. Both the Vapor and NIO teams need to sort it out.

It seems trivial at first, but this situation is actually like attempting to pull a seemingly harmless nail out of the wall in your living room, only to end up having your neighbor's bedroom on the other side of the wall flooded with water and drowning his hamster.

**Comment by Joannis** [comment]
Was this compiled as DEBUG or RELEASE binary? I know that futures are held by SwiftNIO for a while as part of promise leak detection, which indirectly leads to appearant leaks.

**Comment by MihaelIsaev** [comment]
In my case it was compiled in DEBUG

**Comment by Maxim-Inv** [comment]
It compiled as Release binary

**Comment by fouadhatem** [comment]
Release in all tests. I intentionally didn't include debug.

**Comment by fouadhatem** [comment]
@Joannis Just to clarify... All my tests on Linux were tested using both `vapor build` and `vapor run` commands, as well as `swift build -c release` and `swift run --enable-test-discovery` commands. All exhibited the exact same behavior on all configurations, with one exception: `CTRL-C` after `vapor run` gave me back the command prompt at the CLI, while `swift run` did not. It just keeps typing `^C` repeatedly without accepting any further commands.

On the Mac (the one discussed above with relatively lower available resources), Xcode would not be able to terminate the `Run` process when attempting to stop the server, while it was possible to do so every time with `CTRL-C` in a terminal window with both `vapor run` and `swift run --enable-test-discovery`.

**Comment by Joannis** [comment]
Thanks for the clarification all of you! I'll keep you up to date with anything I find.

**Comment by fouadhatem** [comment]
@Joannis There was a slightly more technical discussion about this topic in the NIO channel on Vapor's discord. Here's a link to its most relevant part: https://discord.com/channels/431917998102675485/448584561845338139/766750614994550844

What do you think?

**Comment by fouadhatem** [comment]
Here are the snippets of the discussion from Discord about this issue, as quoted from `@TypeBeta`:

Okay, I think I see something. There's an extremely subtle and obscure way for the signal handler Vapor sets up to get sidetracked in a fashion that prevents it from being handled properly... More CPU cores or memory would definitely affect the occurrence of this little codepath.

You're encountering a particularly interesting bizarre edge case at the intersection between libdispatch's `DispatchQueues` and NIO's `EventLoop`s - Vapor does this to trigger a shutdown on SIGINT (slightly simplified):
```
let promise = context.application.eventLoopGroup.next().makePromise(of: Void.self)
context.application.running = .start(using: promise)
let signalQueue = DispatchQueue(label: "codes.vapor.server.shutdown")
func makeSignalSource(_ code: Int32) {
      let source = DispatchSource.makeSignalSource(signal: code, queue: signalQueue)
      source.setEventHandler { promise.succeed(()) }
      source.resume()
      signal(code, SIG_IGN)
}
makeSignalSource(SIGINT)
```
(several lines snipped, but these are the important bits)

The promise used to actually trigger the shutdown is associated with `application.eventLoopGroup.next()`, in other words "the next event loop that happens to be handy at the moment" - and by default, there are only as many event loops as there are CPU cores.

However, the signal is delivered to and the promise is signaled from a completely separate dispatch queue, which exists totally outside of the event loop system (at least in Vapor's normal usage); NIO event loops are directly bound to threads, but dispatch queues are an entire separate (and one step higher) layer of abstraction. This means that the `process.succeed(())` call must notify the promise's associated event loop by queuing a task on that event loop for execution when the loop can next get to it.

Normally, this is a reasonable thing to do. Promises are used to translate execution between queues and threads in several instances, it's a tested codepath...

But, a dispatch signal source's event handler, while it isn't taking place in the extremely dangerous "async-signal-safe" context of a raw signal handler, is still a side effect of a signal that may well have been originally delivered on the thread belonging to the promise's event loop. The fewer number of threads and event loops there are in general (i.e. the fewer CPU cores and the less RAM the kernel treats as easily spread into thread-specific working set), the more possible it is to subtly "wedge" the promise's event loop so the promise result delivery does not trigger the future result's callback chain - it happens when `EventLoop.inEventLoop` returns an inaccurate result due to the behaviors of dispatch queues. I expounded upon `inEventLoop`'s problems in that regard at quite some length in a [PR](https://github.com/apple/swift-nio-transport-services/pull/84) I got merged into NIO itself that dealt with some somewhat related issues.

The problem normally doesn't arise because signal delivery interrupts an executing thread in a unique fashion that never occurs by any other mechanism.

Of course, the most interesting question is how to work around, avoid, fix, sidestep, or otherwise deal with the problem... That's the part I'm still working on; it's a very messy issue, and it's not possible to (safely) remove either queues or event loops from the logic involved.

**Comment by Lukasa** [comment]
I don't think @gwynne's diagnosis here is right.

Signal sources on Linux are essentially a [wrapper around `signalfd`](https://github.com/apple/swift-corelibs-libdispatch/blob/f13ea5dcc055e5d2d7c02e90d8c9907ca9dc72e1/src/event/event_epoll.c#L169). This means that Dispatch does not do what @gwynne suggests (catch the signal in any random thread and then do a weird self-pipe or something). Instead, what Dispatch does is...much weirder.

The problem with `signalfd` is that [it sucks](https://ldpreload.com/blog/signalfd-is-useless). The specific problem Dispatch has to deal with is that `signalfd` is lower priority than regular signal delivery and so if Linux can find _any_ thread that it can safely deliver the signal to, it'll do that in priority to delivering it to the signalfd (this is not a good behaviour, incidentally).

Dispatch does its best to work around this by, whenever you register a signal source, also installing [a regular signal handler](https://github.com/apple/swift-corelibs-libdispatch/blob/f13ea5dcc055e5d2d7c02e90d8c9907ca9dc72e1/src/event/event_epoll.c#L165). This regular signal handler [catches the signal if delivered to that thread](https://github.com/apple/swift-corelibs-libdispatch/blob/f13ea5dcc055e5d2d7c02e90d8c9907ca9dc72e1/src/event/event_epoll.c#L133), masks off the signal on that thread, and then re-delivers the signal in the hope that this time it'll get delivered to the signalfd instead. This is all reasonably kosher, at least in principle, as `pthread_sigmask` and `pthread_kill` are both signal-safe on Linux.

The upshot here is that whatever reason this is failing is not the result of the interaction of the signal and the event loop. Dispatch queues cannot run on event loop threads, dispatch signal sources always run on dispatch queues, so this is always an across-thread concern. Wherever the issue is, it's elsewhere.

**Comment by fouadhatem** [comment]
@Lukasa Does this in any way apply to macOS as well? I know that Linux and macOS handle signals differently, but for some very elusive (and persistent reason), the behavior seems to be somewhat similar on both platforms, as I have seen. The reason it remains undetected for so long is because (I hypothesize here) most senior devs working on addressing these issues have an abundance of resources on their local machines (CPU cores/RAM). 

I am able to successfully reproduce this issue now on both macOS and Linux whenever there are limited resources on the machines (2 cores or less on both platforms/ 4GB of RAM on Mac - 2GB or less of RAM on Linux). The ONLY exception to all this at the moment is the ability to terminate the server from the CLI on macOS with CTRL-C. However, it fails to terminate on Xcode every single time. The application fails to shutdown, with LLDB citing an RPC server error and a status code of -1.

It took me a while to try and identify a pattern of behavior that could help finally kill this issue once and for all, since there has been a steady stream of reports on the exact same issue going as far back as Vapor 3. 

Here are two issues from earlier this year:

- https://github.com/vapor/vapor/issues/2336

- https://github.com/vapor/vapor/issues/2451

If you like, maybe you can show me how to capture the information you need from my machine to help trace down this issue?

**Comment by Lukasa** [comment]
Yes, all of the analysis applies to macOS as well. macOS doesn't have signalfd so you don't have to deal with the fact that it's bad, and managing signals on macOS is a bit saner and that kqueue is better than signalfd. However, when push comes to shove the signal handlers from DispatchSource still run on non-EventLoop threads, and so there cannot be a concern with them "blocking" an event loop thread. Whatever the issue is, it isn't as described above.

**Comment by Lukasa** [comment]
As to tracing down the issue, the first place I'd look is that I'd look very hard at _how_ Vapor waits for the event loops to be closed.

**Comment by Lukasa** [comment]
The code here is really quite unclear. `Application.start` seems to wait on the future from `running`. The flow control here is really uncertain but assuming that we successfully block there, simply notifying the promise should be sufficient.

I think the right diagnostics to perform are to set breakpoints in `Application.run` to see if `onStop.wait()` actually happens, and to set one in the signal handler. If you don't want to use breakpoints on Linux, you can also just add print statements after each thing we wonder about.

**Comment by gwynne** [comment]
@Lukasa FWIW, I wasn't suggesting that an `EventLoop` was being blocked by a signal handler running on its thread per se - I was speculating (it doesn't deserve any more definite term than that) more that the signal delivery might be happening in such a way as to cause `EventLoop.inEventLoop` to be inaccurate, disrupting the completion of the promise which Vapor uses as a termination signal (due to `SelectableEventLoop.schedule0(_:)` taking the wrong branch).

The behavior of `inEventLoop` is already a problem for `NIOTSEventLoop` in general (see my comments in https://github.com/apple/swift-nio-transport-services/blob/main/Sources/NIOTransportServices/NIOTSEventLoop.swift#L77-L94), though that implementation takes care to fail closed so that tasks take longer to run at worst - the misbehavior I was thinking of in this case (and I must emphasize again I have no evidence other than pure speculation) would not have that safeguard. My (unproven, untested) theory goes that, if `inEventLoop` incorrectly returned `true` when it shouldn't, `SelectableEventLoop.schedule0(_:)` would be misled in such a way that it would _not_ call `self._wakeupSelector()`. In turn, given that the cases of this issue that have been reported have been regarding failure to shut down a server with Ctrl-C (etc.), the odds are high that no further requests to the server in general are likely to arrive - meaning there would also be no external reason for the event loop to be awoken either. End result, callbacks attached to the promise's future result are never run.

If this (unsubstantiated) theory were correct, machines with more CPU cores would have more event loops in the default event loop group - I imagine there would exist a derivation showing that any number of cores greater than a particular threshold could never encounter the necessary conditions to trigger the problem (or the probability falls below some meaningful percentage, etc.). I would further wildly theorize that additional RAM is a red herring, specifically because in all the tests I've seen posted to this issue's discussion thus far, more RAM always also means more CPU cores.

By the way, in case I haven't said it enough times already, this is all totally wild theory, I have no concrete evidence regarding any of it 😆

**Comment by Lukasa** [comment]
It's a wild and impossible theory, sadly, mostly because this prominently happens on Linux. :smile:

On Linux, `inEventLoop` is implemented in terms of `pthread_self` comparisons. This requires that the caller literally be running on the same thread as the event loop. While Dispatch Queues can behave weirdly, it is never possible for a dispatch source callback to execute on a NIO event loop thread on Linux. This is because there are only two places such a call could hypothetically execute. The first option is that it could be `async`'d to the queue and execute on a thread dedicated to that queue: that thread cannot be an event loop thread because it's dedicated to that queue. The other option is it could be `sync`'d or `async`'d to that queue and run on the thread that did the `read` from the `signalfd`. That thread _also_ cannot be an event loop thread, as it is currently issuing a `read` syscall!

In general Dispatch can never arrange to run a block on a `SelectableEventLoop` because to do so it would have to know that `SelectableEventLoop` exists, and it doesn't. There are no false positives for `SelectableEventLoop.inEventLoop`, and so this theory cannot occur. Sadly we have to look closer to home, at how exactly Vapor cleans itself up.

**Comment by fouadhatem** [comment]
@Lukasa what continues to confuse me is the relationship between resources (CPU cores/RAM) and the occurrence of this issue. What parts of Vapor do you think do this? There might be a clue found [here](https://github.com/vapor/vapor/issues/2451), though this isn't dependent on web sockets at all.

**Comment by Lukasa** [comment]
@gwynne is right about the guess that this is likely related to the number of event loop threads.

**Comment by fouadhatem** [comment]
@Lukasa Okay so I am assuming that a fix needs to be done on Vapor then, not NIO, since it is Vapor that handles that. Is this a correct assumption?

**Comment by gwynne** [comment]
Well, in the sense that Vapor is probably the culprit in general, yes. It would not make sense to change the number of event loops, though.

**Comment by fouadhatem** [comment]
@Lukasa @gwynne cool so we seem to have a consensus that this is probably not a NIO issue and most likely a Vapor issue. We do not seem to have a consensus on what on Vapor's side is causing this. 

@Lukasa suspects it might be related to how Vapor waits for it's event loops to close and that it could be related to the number of event loop threads.

@gwynne doesn't think that it might be sensible on Vapor's part to change the number of event loops. I'm guessing you're also leaning towards re-examining how event loops are handled in Vapor as well?

**Comment by mxcl** [comment]
Sorry to spam here, but on the off chance our data point helps:

We use `systemctl` to control our instances and I have noticed that this issue (same as described) only happens if the Vapor process has been running more than a few minutes.

We started seeing it about 5 weeks ago, before that we *were* using Vapor 4 but never saw it. We upgrade our deps about once a week.

**Comment by siemensikkema** [comment]
Let's compare the changes from 4.29.1 and 4.30.0 https://github.com/vapor/vapor/compare/vapor:4.29.1...4.30.0. I see a possible culprit in the way environment files are being loaded. There is a bug there that the EventloopGroupProvider is not forwarded from one `DotEnvFile.load` overload to another. This is a very common bug I see in more and more places. In this case the result of the issue is that it will erroneously create another ELGP. I will open a PR for this one. If this is the cause then the issue should not appear in 4.29.4.

**Comment by siemensikkema** [comment]
@fouadhatem @MihaelIsaev @Maxim-Inv Could any of you test if the issue still persists with 4.32.1?

**Comment by fouadhatem** [comment]
@siemensikkema on it

**Comment by JaapWijnen** [comment]
Yikes, if so this was my fault :/ I see a fix was made to pass in the actual eventloopGroup, but now that I take a second look, I think we also need a fix in the loaddotenv.swift file for when we don't pass an eventloopgroup so that it is shut down after use.

**Comment by JaapWijnen** [comment]
Ah no, double checked, it should shut down the eventLoopGroup in case of a `.createNew` `EventLoopGroupProvider`

**Comment by fouadhatem** [comment]
@siemensikkema I think you're on the right path. Preliminary testing on my end shows successful server shutdown on some instance configurations, but not all. It still fails on a limited capacity MacBook Air and a few other Linux instances. I am continuing to test as I type this. Will report in more detail once I have the full picture. However, it seems to me that if this is a widespread issue across multiple areas of Vapor, then I think that this issue would be at the very least partially resolved, if not entirely, if similar bugs are dealt with across the project.

@JaapWijnen this seems to be most definitely related to what @Lukasa mentioned earlier regarding how event loops are closed and what might cause this process to be interrupted, as event loops seem to have a strong correlation with narrowing availability of system resources.

**Comment by JaapWijnen** [comment]
Yeah, I just missed that I wasn't passing the eventLoopGroupProvider to the function that does the actual dotEnvFileLoading so it was creating two new `EventLoopGroup`s before 4.30.1  (one for every .env file) 😓Fixed now by #2514 luckily. Shouldn't be a problem in itself indeed, but shouldn't have happened ofc.

**Comment by fouadhatem** [comment]
Don't sweat it. That wasn't the root cause of it, but it did make it a little worse. However, it seems that this is what might have led @siemensikkema to detect the pattern of errors causing it in the first place. So thanks! :)

**Comment by fouadhatem** [comment]
@siemensikkema also, you are right in that the issue did not appear on Linux on on 4.29.4 and before. However, 4.28 fixed an issue that caused a similar effect but because of an entirely different reason (open web sockets as a result of browser keep-alive). 

However, it seemed to sporadically appear on Xcode with smaller capacities (MacBook Air 2013 with 4GB RAM in my case) throughout this time and it only seemed to get worse with every new tag, until it entirely stopped terminating successfully on Xcode, always complaining about:
```
Message from debugger: The LLDB RPC server has exited unexpectedly. Please file a bug if you have reproducible steps.
Program ended with exit code: -1
```
This may be happening for a different reason, but the previous randomness of its occurrence always seemed to point to resource availability issues. For example, if I had more open programs running in the background, it would fail to terminate, leading me initially to believe that this may be an issue of memory leaks.

**Comment by fouadhatem** [comment]
@siemensikkema Here's more info on how the issue responds to 4.32.1 as compared to 4.32.0:

(Y = Successful shutdown, N = Unsuccessful shutdown)

**Linux (Ubuntu 20.04 on VMWare):**
- 1 CPU Core - 1GB RAM:  **4.32.0** = N, **4.32.1** = N
- 1 CPU Core - 2GB RAM:  **4.32.0** = N, **4.32.1** = N
- 2 CPU Cores - 1GB RAM:  **4.32.0** = N, **4.32.1** = Y
- 2 CPU Cores - 2GB RAM:  **4.32.0** = N*, **4.32.1** = Y  |  *The 4.32.0 instance produced the following error when attempting to shutdown:
```
ERROR: Cannot schedule tasks on an EventLoop that has already shutdown. This will be upgraded to a forced crash in future SwiftNIO versions.
```
- 2 CPU Cores - 4GB RAM: **4.32.0** = Y, **4.32.1** = Y

**Mac (macOS 10.15.7 and Xcode 12.1):**
- MacBook Air (2013) with 2 CPU Cores(i5) - 4GB RAM: **4.32.0** = N, **4.32.1** = N
- MacBook Pro (2019) with 6 CPU Cores(i7) - 16GB RAM: **4.32.0** = Y, **4.32.1** = Y

**Comment by siemensikkema** [comment]
@fouadhatem I found some more likely culprits which are now fixed and deployed and available in 4.34.0. Would you mind having another look?

**Comment by fouadhatem** [comment]
@siemensikkema on it

**Comment by fouadhatem** [comment]
@siemensikkema Here's more info on how the issue responds to 4.34.0 as compared to 4.32.1:

(Y = Successful shutdown, N = Unsuccessful shutdown)

-----

**Linux (Ubuntu 20.04 on VMWare):**
- 1 CPU Core - 1GB RAM:  **4.32.1** = N, **4.34.0** = Y   **( Improvement )**
- 1 CPU Core - 2GB RAM:  **4.32.1** = N, **4.34.0** = Y  **( Improvement )**
- 2 CPU Cores - 1GB RAM:  **4.32.1** = Y, **4.34.0** = Y  **( No change )**
- 2 CPU Cores - 2GB RAM:  **4.32.1** = Y, **4.34.0** = Y **( No change )**
- 2 CPU Cores - 4GB RAM: **4.32.1** = Y, **4.34.0** = Y **( No change )**

All VMWare tests passed. Testing may be required to confirm success on cloud-based deployments. @Maxim-Inv @mxcl @MihaelIsaev @jhoughjr can you kindly test and confirm these findings on your resources? Particularly, @mxcl can you check if the issue persists if you run Vapor instances for more than 10 minutes?

-----

**Mac (macOS 10.15.7 and Xcode 12.1):**
- MacBook Air (2013) with 2 CPU Cores(i5) - 4GB RAM: **4.32.1** = N, **4.34.0** = N  **( No change )**
- MacBook Pro (2019) with 6 CPU Cores(i7) - 16GB RAM: **4.32.1** = Y*, **4.34.0** = Y*  **( No change )**

*The MacBook Pro displays random intermittent failures to shutdown the server, particularly on the very first run of a project, or when attempting to run the project for the first time after a system reboot. This happens an average of 2 times for every 10 runs of the same unchanged project (assuming one system reboot).

This is the same pattern that was first observed on the MacBook Air when the issue first started presenting itself ever since I started using Vapor 4.

@mikecsh can you try out 4.34.0 on your MacBook and confirm what you see?

-----

This overall pattern continues to confirm that this is indeed an issue directly relating to incorrect creation/closing of different event loops across the platform, as shown by 4.34.0 which quashed a few more of these bugs. 

I continue to note that the issue on macOS could be (but as of now is not yet definitely) unrelated to the cause of these issues on Linux (erroneous creation/closing of event loops). Advice is needed on extracting information from Xcode to identify the source of the issue and confirm whether it is indeed related to the above.

**Comment by MihaelIsaev** [comment]
I've just tested `4.34.0` on Ubuntu 20.04 on DigitalOcean and `Ctrl+C` seems working now.

Droplet config: Shared 1 vCPU 1GB RAM $5/mo

**Comment by siemensikkema** [comment]
@fouadhatem thanks for the detailed analysis! So we can conclude that the issue is reduced to happening only on Xcode on macOS and that this is not a recently introduced issue? Would it be outlandish to suspect if Xcode itself might the problem?

**Comment by MihaelIsaev** [comment]
@fouadhatem Could you please try to run your app using just `swift run` and try to press `Ctrl+C` in terminal on your `MacBook Air (2013) with 2 CPU Cores(i5) - 4GB RAM` to exclude possible problems with Xcode

**Comment by siemensikkema** [comment]
@MihaelIsaev if I understand correctly, that behavior has been working fine:
> The MacBooks continue to be able to build, run and terminate any Vapor 4 project successfully from the Terminal.

**Comment by MihaelIsaev** [comment]
Sorry, I missed that part 🙂

**Comment by fouadhatem** [comment]
@MihaelIsaev projects build, run and terminate just fine on the MacBook Air using both `vapor run` and `swift run`

@siemensikkema about Xcode, it certainly is a possibility that Xcode could be the culprit here, though I'm not sure why. 

One pattern of behavior that I did not mentioned before, which may indicate Xcode as being the source of the problem, is the fact that earlier tags of Vapor that were able to terminate successfully most of the time on earlier releases of Xcode. With 12.1, this is no longer possible. Projects consistently fail to terminate on the MacBook Air. I can investigate this by downloading an earlier release of Xcode and experimenting on 4.34.0 to see if it terminates. If this works, then we can start attempting to nail down why it happens.

I'm not very well-versed with the inner workings of Xcode to be certain. I would be happy to follow any instructions on how to dig deeper into Xcode to investigate this, which incidentally may spare this poor MacBook Air the torture of multiple downloads and installs of previous versions of Xcode.

On another not, potential culprits similar to the ones you have been locating so far in Vapor that relate to this issue may be found in the following files:

- `Sources/Vapor/HTTP/Headers/HTTPHeaders+Connection.swift`

- `Sources/Vapor/HTTP/Server/HTTPServer.swift`

- `Sources/Vapor/HTTP/Server/HTTPServerHandler.swift`

- `Sources/Vapor/HTTP/Server/HTTPServerRequestDecoder.swift`

- `Tests/VaporTests/ServerTests.swift`

- `Tests/VaporTests/WebSocketTests.swift`

The reason I suspect them is because these are the files that were changed as part of 4.28.0. They directly control HOW a server is terminated, rather than the indirect impact of erroneous event loops that you have been finding so far. Would you kindly take a look?

**Comment by gwynne** [comment]
Might be related to the weird change in `SIGCONT` behavior in LLDB. See also https://twitter.com/stuartcarnie/status/1316765381923491847

**Comment by gwynne** [comment]
Nothing like hitting the close issue button by accident 😓

**Comment by fouadhatem** [comment]
@gwynne ye who wieldith thy powereth shall deployeth it wisely..eth

And how the hell did you find that `SIGCONT` thing?? It feels relevant because it's always LLDB that complains about an RPC server quitting unexpectedly when I attempt to terminate any Vapor project on Xcode on that downtrodden MacBook Air of mine. 

Do you have a theory as to why it might not be an issue on more powerful macs, while still being an issue on less resourceful ones?

**Comment by mxcl** [comment]
Didn’t seem to help us either unfortunately.

**Comment by mikecsh** [comment]
@fouadhatem Thanks for the prod - I'm still seeing the issue with 4.34.0 on my 12" MacBook both under docker and XCode.

**Comment by fouadhatem** [comment]
@mikecsh can you remind me please what the specs of your machine are, particularly processor type, number of cores, RAM and which version of macOS/Xcode you're running?

**Comment by mikecsh** [comment]
@fouadhatem 

Xcode 12.0.1  and Xcode 12.2 beta-3 and Ubuntu 18.04 under Docker (Engine v19.03.13) on a 2017 1.4 GHz Dual-Core Intel Core i7 MacBook 12" with 8 GB RAM running macOS 11.0 Beta (20A5395g).

This was occurring before upgrading to the macOS 11 Beta also on 10.15.4.

**Comment by fouadhatem** [comment]
@mxcl the reason I mentioned this issue to you was the randomness of the occurrences you are facing, which is very similar to what we are facing here. 

Identifying the nature of the bug here would very probably help us identify the nature of the bug in your issue with HTTP2, or vice versa. The pattern of behavior in both cases is too similar to ignore (failure to reliably reproduce, randomness).

**Comment by fouadhatem** [comment]
@mikecsh Thanks.

Just so I'm clear: you're running Docker and Ubuntu 18.04 on the same machine?

**Comment by mikecsh** [comment]
@fouadhatem  Sorry, that wasn't too clear.

I tend to develop, build, and test my Vapor app under Xcode/macOS for a quicker dev cycle due to very slow compile times under Linux

After each significant stage of development I update a version of it running under Ubuntu 18.04 inside Docker (on the same machine) where the Vapor app is integrated with other components of my system.

After that it is deployed to a $5 DO droplet. I have no data points from there as that is still running a version based on Vapor v3

**Comment by fouadhatem** [comment]
@mikecsh okay that makes more sense. So if I understand this correctly, you still face the issue when attempting to compile and run on Xcode during your dev phase?

**Comment by mikecsh** [comment]
@fouadhatem Yes that is what I was seeing earlier, however I've just tried to reproduce it again (in Xcode 12.2 Beta-3) and I'm not able to. I have tried both terminating within a few seconds of launching and having let it run for 10 minutes. I'll report back if I see the issue again under Xcode.

I am still seeing it under Docker/Ubuntu 18.04.

Thanks for looking into this.

**Comment by fouadhatem** [comment]
@mikecsh my pleasure! It affects me as well so I kind of have to as well lol.

This is very interesting! Xcode 12.2 Beta-3 does NOT show the issue, but earlier (stable) versions of Xcode do.

As for Ubuntu and Docker, can you please tell me which packages you have imported from Vapor into your project? Also, can you run an empty Vapor template project with the same set up and tell me if you still see the issue?

Thanks for helping out!

**Comment by fouadhatem** [comment]
@siemensikkema @mikecsh okay we have a mysterious development:

On the MacBook Air(dual core i5 with 4GB RAM), Xcode fails to terminate the project after the very first run. However, after the I manually kill the `Run` process and run the same unchanged project again from Xcode, it manages to terminate successfully every time. The pattern repeats if I quit and reopen Xcode or if I reboot the laptop. I will try to reproduce this in different scenarios.

Have any of the core vapor/fluent packages been updated recently?

**Comment by fouadhatem** [comment]
@siemensikkema detected an interesting new pattern relating to this issue:

If an Xcode project crashes at runtime for any reason during its very first run, the issue disappears for as long as that Xcode project is open, even if it is being changed and rebuilt. However, if you quit Xcode and relaunch it again with the same project and attempt to run it, the failure to terminate pattern will return.

Example: I create an Xcode project from template with the usual packages(Vapor, Fluent, Fluent Postgres Driver) and I create a couple of models and their associated migrations. If I deliberately mess up a migration (reference a non-existent parent model from a child model, for instance) then the error will only be triggered at runtime when Vapor attempts to run the migration and the server will not start. 

Fixing the references in the migration, resetting the db and attempting to run it again will cause the project to terminate correctly each and every time thereafter. However, if I quit Xcode and attempt to run the project again, the same old pattern of unsuccessful server shutdown returns.

**Comment by gwynne** [comment]
@fouadhatem Yep, even more definitely sounds specifically related to LLDB - Can you try editing the scheme and unchecking the "Debug executable" box in the Info tab of the Run action, and see if the issue remains? Preferably using Xcode 12.2b4, in particular. (In fact, it'd be useful to know whether the issue is present in that Xcode version at all.)

**Comment by fouadhatem** [comment]
Oh wow you DO levitate! @gwynne I think we have a winner! Works every time with `Debug Executable` box unchecked. 

This is on Xcode 12.1 and I'm downloading 12.2b4 to further test out this theory. I'm also going to try out different configurations for the settings below to see which other combinations cause the same issue to arise and which make it go away.

My schema settings (under Run / Info) for the project are now:
- Build Configuration: Debug
- Executable: Run
- Debug Executable: Unchecked
- Debug process as: Me
- LLDB Init File: `$(SRCROOT)/LLDBInitFile`
- Launch: Automatically

How did you figure this out and what is causing it?

**Comment by gwynne** [comment]
@fouadhatem When combined with the `SIGCONT` issue I referenced earlier in this discussion (https://github.com/vapor/vapor/issues/2502#issuecomment-717525115), the issue going away after the first failed termination but returning when Xcode was relaunched was a dead giveaway - relaunching Xcode also relaunches the macOS-specific `lldb-rpc-server`, and given that termination is implemented by delivery of a signal (`SIGINT` or `SIGKILL`, depending on termination method), the existing LLDB signal handling issue was an obvious culprit.

Of course, this has nothing whatsoever to do with any issues experienced on Linux, but it looks like the event loop fixes actually solved that, and the LLDB issue was an unfortunately well-timed coincidence.

**Comment by fouadhatem** [comment]
@gwynne very interesting! Why did you ask me about Xcode 12.2b4 though? Is that LLDB issue resolved there? There was a reference in this thread earlier about the issue not popping up in Xcode 12.2b3.

**Comment by gwynne** [comment]
@fouadhatem Yeah, the mention of it vanishing in b3 made me wonder if it'd appear at all in b4.

The issue with LLDB is one of mishandled signal delivery, caused by - as far as I can tell - a bug in the RPC-based interface with Xcode (all of which is internal to Apple's developer tools, there's no public RPC mode for LLDB as far as I'm aware). It resulted in LLDB treating certain signals (such as `SIGCONT`) as breakpoints, rather than using the "pass to target process" behavior those signals previously had. My best guess is that since Xcode was issuing a termination command, the UI wouldn't stop at a breakpoint for it, but LLDB would still fail to pass the appropriate signal along, so the termination wouldn't happen. On next launch, the signal handling behavior would be reset to the correct defaults (since the mixup seemed to be due to Xcode's internal one-time initialization after launching the RPC server process), and everything would work - until the RPC server got relaunched and Xcode ran its own initialization again.

I would guess that the "only when running with constrained resources" behavior was already absent once the event loop issues were fixed and the bug vanished completely on Linux - but of course, the LLDB bug remained, and it would take some careful re-examination to realize that the issue was suddenly affecting _all_ systems.

**Comment by fouadhatem** [comment]
@gwynne thanks for that!

I just tried different combinations of settings and the issue only completely goes away when the `Debug Executable` box is unchecked. This indeed was a very unfortunately well-timed coincidence! How about that?

**Comment by fouadhatem** [comment]
@mikecsh can you implement the solution from above and let us know if that resolved it for you on the Mac on your non-beta Xcode version?

Also, are you still experiencing the issue on Docker with Ubuntu 18?

**Comment by fouadhatem** [comment]
@gwynne Assuming that everyone else who participated in this issue are now satisfied with the resolutions mentioned here for all platforms, do you think it needs to be extended anywhere else after this issue is closed to make sure that the findings here are not hard to find in the future?

Something along the lines of:

- Adding some kind of tip or note somewhere in [here](https://docs.vapor.codes/4.0/xcode/) that talks about turning off that `Debug Executable` option in case someone else is facing a similar issue when using Xcode versions lower than 12.2, followed by a small warning regarding what info/functionality could be lost when that option is turned off.

- Filing a bug report with Apple and/or LLVM about this bug.

Speaking of which, is it really an LLDB bug or is it more of an Xcode thing? The reason I ask is that I never had issues building/running/terminating a Vapor project from the macOS terminal, which I assume also uses the same LLVM/LLDB infrastructure. Or am I missing something else, as usual?

**Comment by admkopec** [comment]
> * Filing a bug report with Apple and/or LLVM about this bug.

@fouadhatem I have filed a bug report with Apple back in summer when Xcode 12 was still in beta (the issue is not occurring on Xcode 11). I didn’t put much effort into it, due to the lack of crash logs generated, however more recently I was able to find a connection between this issue and the macOS deployment target. The default value for swift package is to use macOS 10.10, and that target is used by Vapor 3, Vapor 4 however sets this target to macOS 10.15 and I don’t get this “crash” on my MacBook Pro.

> Speaking of which, is it really an LLDB bug or is it more of an Xcode thing? The reason I ask is that I never had issues building/running/terminating a Vapor project from the macOS terminal, which I assume also uses the same LLVM/LLDB infrastructure. Or am I missing something else, as usual?

Xcode is using LLDB in a slightly more advanced way, maybe even a different version 🤔, than the terminal, to allow for the UI debugging features, etc.

> _**EDIT:**_ Just tried it out on Xcode 12.2. The issue still persists and it looks like your solution is the only one that fixes it.

The other solution is to try to “bump” the deployment target in Xcode for macOS to at least 10.11 for Vapor 3 (I am unable to reproduce this issue on Vapor 4, as its deployment target is already 10.15). Otherwise the issue is still present both in Xcode 12.3 and 12.4

**Comment by fouadhatem** [comment]
@admkopec thanks for this!

Can you please elaborate a bit on the relationship between the macOS deployment target, LLVM infrastructure and this issue? It would be very interesting to know where they intersect.

**Comment by admkopec** [comment]
@fouadhatem Firstly I want to mention that my research is limited to Vapor 3, as I couldn't replicate the issue on Vapor 4, and there have been multiple changes made in the version 4, so it might not fully explain the origin of this issue.

At first I suspected the issue might be some kind of a timeout being triggered by the Vapor's `Application` class, as when this class is initialised and allocated at the time when LLDB tries to terminate the process, the debugged process hangs and after a few seconds the debugger throws an error/crashes. At this point the process returns to normal operation - it is not terminated. However when we deallocate the `Application` class before we try to terminate the process _(for example if we put it only onto stack as a function only variable, instead of a global var and the function completes execution)_, everything is fine - no error is thrown. This also happens even if the `NIOServer` is not in the running state, only the `Application` class has to be initialised.

As I was conducting these tests on a large AppKit based macOS target, I decided to make a bare minimum Vapor executable based off of the api-template for Vapor 3. Soon enough I was still able to consistently reproduce this hang and then crash behaviour even on a very lightweight command line executable. Then I tried to replicate this behaviour on Vapor 4 - but I wasn't able to. And the first thing that caught my attention was the deployment target set on Vapor 4. Once I updated the `Package.swift` file of my lightweight api-template to use the same deployment target - macOS 10.15, the issue was gone. The process instantly terminates and gives the `exit code: 9`. Then I tried to check every deployment target between 10.10 and 10.15 to try and replicate the issue, **only the deployment target set to 10.10 hangs and results in a crash of the debugger**. But I also think that at the macOS 10.11 target the time between clicking the Stop button and terminating the process was slightly longer than the deployment target of 10.15, _but I would have to check it using some kind of automated tests or measurement utility._

Then with these finding I got back to my AppKit app and after changing the deployment target to macOS 10.11 for Vapor framework _(it doesn't matter if Vapor is linked statically or dynamically)_ the issue was no longer occurring, though the time between clicking the Stop button  and termination was significantly longer than the api-template, around the time the hang takes when the issue is occurring. _(I also don't get any exit codes, but I don't remember now if I ever did.)_

However at one point, _and I couldn't replicate it yet_, after clicking the stop button on this larger app, the debugger exited successfully, it didn't throw any errors or crashes, but the app still **didn't terminate**.

So in conclusion, I still cannot prove nor disprove if this issue is some kind of a timeout or not, I suspect that because the binary is slightly smaller when the deployment target is set for a newer version of macOS (better optimisations, etc.) it might explain why the debugger wouldn't timeout then, but it doesn't explain the issue I mentioned, when the app didn't terminate, yet the debugger exited successfully.

I will attach my sample project here, so if anyone wants to experiment feel free to do so, I also shared my finding and the project with Apple as part of my report, if anyone has been able to get the LLDB to put a crash log into the `~/Library/Logs/DiagnosticReports` directory, let me know, so that I could forward it to Apple and maybe get some more info about this problem.
[api-template.zip](https://github.com/vapor/vapor/files/5932413/api-template.zip)

**Comment by fouadhatem** [comment]
@admkopec sounds quite comprehensive! I'll put my chips in with yours on the issue being somewhere in the LLDB. 

Where do you think the RPC server fits into all this though?

**Comment by admkopec** [comment]
@fouadhatem I suspect, as far as I know, the RPC server is used for some kind of communication between processes, so that could be the first part to crash when something starts breaking. What concerns me the most however is that there are no crash logs generated when the issue occurs. Just as though everything went smoothly and nothing crashed, except from the "message from debugger".

Additionally I consistently get these messages in my Console logs (deployment target macOS 10.10):
<img width="1774" alt="Screenshot 2021-02-05 at 23 55 32" src="https://user-images.githubusercontent.com/14315425/107098345-4b2d7300-680f-11eb-8a7f-24b3601337d7.png">
When I however change the config to anywhere from macOS 10.11 to 10.15, I get less messages, but it actually now appears to be sending some PTKILL to the debugged process, there are also no messages whatsoever from the RPC server:
<img width="1774" alt="Screenshot 2021-02-05 at 23 56 44" src="https://user-images.githubusercontent.com/14315425/107098410-6bf5c880-680f-11eb-95d5-05c34de08108.png">

For some reason, when the issue occurs, `ptrace` is asked to pass signal **-1**, which appears to be invalid and that breaks the debug server, which could manifest itself through the RPC server crash. It might appear as if this signal is some kind of placeholder value to use when there is no precise value set, (like after timeout?), but I still don't know where this signal comes from, Xcode, RPC server, somewhere else? Also I don't know why the RPC server manifests its presence only when the deployment target is set to macOS 10.10, but that sheds some light why I didn't get the RPC server crash when my deployment target was set to macOS 10.11, and yet the process still didn't terminate.

**Comment by fouadhatem** [comment]
@admkopec I'm curious about something. Would you please try out the following steps and let me know what happens on Vapor 3:

1. Create a project from scratch using your template while setting the deployment target to 10.10. This should cause your application to crash once you build and run and you should receive a `-1` exit code on the console in Xcode. Keep Xcode open.
2. Open activity monitor and make sure that both the `Run` process and the` lldb-rpc-server` process are still running. They should still be there since you received a `-1` exit code. It might be easier to switch to the Memory tab in activity monitor to keep an eye on the latter process memory usage in particular.
3. Without killing either of the two processes (keep them both running), build and run the project again on Xcode. That should reliably fail because you would still have the `Run` process hijacking the target port. What exit code do you see here?
4. Still without closing Xcode, manually kill the `Run` process and then build and run the project once more. Do you still get a `-1` here?

**Comment by admkopec** [comment]
@fouadhatem After getting the `-1` exit code the `Run` process is running, but the `lldb-rpc-server` in my case terminates, then after trying to build and run again I get a `fatalError` on `Environment.detect()` about not being able to bind to the port (Address already in use), which is expected as the `Run` process is still running. Then once I terminate from Xcode I get exit code `9`, Application class wasn't yet initialised. Although when I terminate at that point `lldb-rpc-server` is still running, but on subsequent build and runs I am still getting the same `fatalError` and exit code `9`. However when I now terminate the `Run` process and let `lldb-rpc-server` keep running, on additional builds I keep consistently getting exit code `9` without crashes and `Run` process terminates, while `lldb-rpc-server` is still running.

However when I quit the `Run` process after the `-1` exit code  and change the deployment target to macOS 10.15, without closing Xcode or project, on the first run I still get the `-1` exit code and `lldb-rpc-server` crash, but on second and subsequent builds, either way when I terminate the `Run` process, or not, I am getting exit code `9` and `lldb-rpc-server` is not terminating.

When I then change back to macOS 10.10, right on the first run I get exit code `-1` and `lldb-rpc-server` terminates, while `Run` process is still running. However, now I get slightly different set of messages in Console, although the -1 signal is still present.
<img width="1774" alt="Screenshot 2021-02-06 at 09 58 08" src="https://user-images.githubusercontent.com/14315425/107113863-0121ad00-6862-11eb-8592-7fe0c4a76b6b.png">

**Comment by gwynne** [comment]
While changing which version of the macOS SDK is being used to build an app will change all kinds of behaviors in `libSystem`, `dyld`, `Foundation`, `AppKit`, etc., these are "linked on or after" checks. Changing just the deployment target is just an `Info.plist` key, though, and usually doesn't affect system framework behaviors nearly as much (though there are still some). It's possible that LLDB, `lldb-rpc-server`, or some other related piece is sensitive to a change made somewhere between Yosemite and Catalina, and this is why the behavior differs slightly. 

The error you're seeing there is LLDB receiving an `EINVAL` error (as it says in the log, "invalid argument", in this case a nonexistent thread ID) in response to its attempt to use `PT_THUPDATE` to explicitly specify the signal to be delivered to the thread. Making this call is part of LLDB's normal operations - signal interception and manipulation on macOS is implemented by catching the `EXC_SOFTWARE`/`EXC_SOFT_SIGNAL` Mach exception type and subtype to intercept a signal before it goes through the BSD signal machinery, then using `ptrace(PT_THUPDATE)` to specify what signal to actually deliver to the thread (if it's not the same one that originally came in, the original is suppressed and replaced with the new one).

What it looks like is happening in this case is LLDB was too slow on the uptake, and the specific thread had already exited (and thus invalidated its ID) between the time when LLDB intercepted the Mach exception and when it issued the `ptrace()` call. The signal LLDB intended to deliver (presumably signal 9 - `SIGKILL`) is thus never delivered, and the process doesn't exit.

**Comment by gwynne** [comment]
tl;dr: ~Thread~ Process management is extremely difficult and error-prone.

(How much of this is unavoidable given the general complexity of efficiently scheduling tasks on multiprocessing systems, and how much is the crushing weight of some fifty years or so of legacy API designs piled on top of each other and forcing obedience to semantics that haven't been optimal choices since back when 16 bits was the standard machine word size on a cutting-edge top-of-the-line CPU... well, that's up for debate. Basically, I blame Intel for a lot of the terrible state both BSD and POSIX are in to this day.)

**Comment by fouadhatem** [comment]
@admkopec Okay so we seem to have somewhat figured out a way to replicate this, and @gwynne 's analysis explains it. 

@admkopec I'm curious; what are the specs of your machine?

**Comment by admkopec** [comment]
@gwynne Thanks for the info about `ptrace` and signaling, I have checked the behaviour both on Catalina and Big Sur running Xcode 12 and the issue is present on both OSes, can't say much about Console logs though if they differ or not, but the issue didn't exist on Catalina running Xcode 11, so I'm not sure if it's really a kernel problem, as it should also manifest itself with Xcode 11? But simply opening the project in the previous version of Xcode "fixed" the issue, no other changes required. I also believe the change, you mentioned, could have been made as far back as between Yosemite and El Capitan, as that's the first version that seems to quite reliably handle the debugged process termination on Xcode 12.

@fouadhatem As for the specs, I have a 2017 15" MacBook Pro, so it's a 2.9GHz Quad-core i7 and 16GB of RAM.

**Comment by gwynne** [comment]
Yeah, I'm sticking with the theory that it's a bug in LLDB specifically, then. That's the easiest way to explain it being affected by different Xcode versions.

**Comment by fouadhatem** [comment]
@gwynne As far as you know, is there an LLDB bug report that has been filed regarding this issue? And do you think that @admkopec 's issue is related to this one?

@admkopec thanks for all your input!

**Comment by fouadhatem** [comment]
And yeah... Intel sucks.

**Comment by mxcl** [comment]
So I feel this thread was somewhat derailed by blaming LLDB.

I get this in production relatively often, and LLDB isn't even installed there. Just now I got:

```
May 11 18:24:11 ip-172-31-90-104 systemd[1]: Stopping "Caché Vapor API"...
May 11 18:25:41 ip-172-31-90-104 systemd[1]: api.service: State 'stop-sigterm' timed out. Killing.
May 11 18:25:41 ip-172-31-90-104 systemd[1]: api.service: Killing process 422620 (serve) with signal SIGKILL.
May 11 18:25:41 ip-172-31-90-104 systemd[1]: api.service: Killing process 422621 (NIO-ELT-0-#0) with signal SIGKILL.
May 11 18:25:41 ip-172-31-90-104 systemd[1]: api.service: Killing process 422622 (NIO-ELT-0-#1) with signal SIGKILL.
May 11 18:25:41 ip-172-31-90-104 systemd[1]: api.service: Killing process 422623 (TP-#0) with signal SIGKILL.
May 11 18:25:41 ip-172-31-90-104 systemd[1]: api.service: Killing process 422624 (serve) with signal SIGKILL.
May 11 18:25:41 ip-172-31-90-104 systemd[1]: api.service: Main process exited, code=killed, status=9/KILL
May 11 18:25:41 ip-172-31-90-104 systemd[1]: api.service: Failed with result 'timeout'.
May 11 18:25:41 ip-172-31-90-104 systemd[1]: Stopped "Caché Vapor API".
May 11 18:25:52 ip-172-31-90-104 systemd[1]: Started "Caché Vapor API".
```

After 90s my systemd kills it waiting.

The thread is quite hard to read now, so apologies for not helping and just reopening it.

I have had this issue on a 16GB machine and a 2GB RAM one also. Both Ubuntu 18.04 running a raw binary, no docker.

**Comment by fouadhatem** [comment]
@mkll I don't think it's strictly an OS problem. I'm also not sure if SwiftNIO or Vapor are the culprits here, though I suspect that one of them could be. 

How does your implementation handle signals and - more interestingly - what made you write your own signal handlers in the first place? Did you face something similar to this issue?

**Comment by 0xTim** [comment]
Thank you all for continuing to raise this. I was under the impression that things were pretty much fixed apart from the Xcode bug, but looks like that isn't the case. 

@mxcl are you doing anything with your own signal interrupts, Swift Service Lifecycle or something that uses Swift Service Lifecycle (like a metrics package etc). Or do you have any long running processes that would prevent a SIGTERM from completing? I'll do a review tomorrow and see if I can spot anything and I'll try and reproduce

**Comment by mkll** [comment]
@fouadhatem Yes, I faced loss of signal in Vapor, after which it was killed by systemd. Since this happened about 20-30% of the time, I lost my temper pretty quickly. ;)

As for the rest of your questions, I believe the source code will be a comprehensive answer, so I placed it [here](https://gist.github.com/mkll/45ed6dd5a773c896f67ddfe999fc9aef). Basically, there is nothing special there, just an adaptation of a similar handler from C.

**Comment by mkll** [comment]
@fouadhatem  I don't think this solution _will_ fix it for really everyone, but symptoms was the same, so it _may_ fix. Just try.

**Comment by mxcl** [comment]
@fouadhatem trying it out, thanks all (esp. @mkll)

**Comment by lmcd** [comment]
Is this issue anything to do with `app.server.shutdown()` not working?
It's never worked for me, and I only ever see `ERROR HTTPServer.shutdown():291 - Could not stop HTTP server: Abort.500: Server stop took too long.`

Seems to be getting stuck in `ServerQuiescingHelper`

**Comment by fouadhatem** [comment]
It's possible. Do you get any LLDB errors as well?

**Comment by lmcd** [comment]
@fouadhatem - no, not that I can see.
Running on macOS - HTTP/2. `shutdown` called outside of an event loop on the main thread.
Worth noting that the server does stop accepting connections, but things like `WebSocket` connections aren't properly closed.

**Comment by ryanfrancesconi** [comment]
I see this issue in vapor 4.65.2 running the server as a localhost on macOS 12 with websockets from a browser or other client connecting to it. The server will shut down fine as long as no websockets have connected. It's possible I'm doing something wrong as I'm just attempting to evaluate vapor for a project, but without the ability to cleanly shutdown and restart, I can't use it - so it's difficult to evaluate. I'm also running the server in the background via AppKit and trigger the shutdown with a cancel. 

Here's my experience with one socket connected:

```
[ codes.vapor.application ] [ INFO ] run() (SocketServerTestHarness/SocketServerController.swift:135)
[ codes.vapor.application ] [ NOTICE ] Server starting on http://127.0.0.1:8080 (Vapor/HTTPServer.swift:296)
[ codes.vapor.application ] [ TRACE ] Decoded HTTP part: head: HTTPRequestHead { method: GET, uri: "/add/socket", version: HTTP/1.1, headers: [("Host", "127.0.0.1:8080"), ("Connection", "Upgrade"), ("Pragma", "no-cache"), ("Cache-Control", "no-cache"), ("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 12_5_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36"), ("Upgrade", "websocket"), ("Origin", "file://"), ("Sec-WebSocket-Version", "13"), ("Accept-Encoding", "gzip, deflate, br"), ("Accept-Language", "en-US,en;q=0.9"), ("Sec-WebSocket-Key", "sPbU3mxJmHYqNArW3OrtWg=="), ("Sec-WebSocket-Extensions", "permessage-deflate; client_max_window_bits")] } (Vapor/HTTPServerRequestDecoder.swift:34)
[ codes.vapor.application ] [ TRACE ] Decoded HTTP part: end (Vapor/HTTPServerRequestDecoder.swift:34)
[ codes.vapor.application ] [ INFO ] GET /add/socket [request-id: 60EACB95-6B50-4800-A7FC-95B510DEC6AB] (Vapor/RouteLoggingMiddleware.swift:13)
[ codes.vapor.application ] [ TRACE ] Response already sent, skipping request body. (Vapor/HTTPServerRequestDecoder.swift:201)
2022-09-20T08:07:07-0700 info SocketServer : [SocketServerTestHarness] Sending: Response(id: 80068C85-2AD8-46C8-9C53-C86D899F839D, type: handshake, string: hello)
2022-09-20T08:07:07-0700 info SocketServer : [SocketServerTestHarness] Sending SocketServerResponse to socket(WebSocketKit.WebSocket)
2022-09-20T08:07:20-0700 info 127.0.0.1:8080 : [SocketServerTestHarness] cancelling task...
[ codes.vapor.application ] [ DEBUG ] cancel handler called (SocketServerTestHarness/SocketServerController.swift:128)
[ codes.vapor.application ] [ DEBUG ] Application shutting down (Vapor/Application.swift:135)
[ codes.vapor.application ] [ TRACE ] Shutting down providers (Vapor/Application.swift:137)
[ codes.vapor.application ] [ TRACE ] Clearing Application storage (Vapor/Application.swift:141)
[ codes.vapor.application ] [ DEBUG ] Requesting HTTP server shutdown (Vapor/HTTPServer.swift:314)
[ codes.vapor.application ] [ ERROR ] Could not stop HTTP server: Abort.500: Server stop took too long. (Vapor/HTTPServer.swift:318)
[ codes.vapor.application ] [ DEBUG ] HTTP server shutting down (Vapor/HTTPServer.swift:320)
[ codes.vapor.application ] [ TRACE ] Shutting down EventLoopGroup (Vapor/Application.swift:149)
ERROR: Cannot schedule tasks on an EventLoop that has already shut down. This will be upgraded to a forced crash in future SwiftNIO versions.
[ codes.vapor.application ] [ TRACE ] Application shutdown complete (Vapor/Application.swift:158)
2022-09-20T08:07:30-0700 info 127.0.0.1:8080 : [SocketServerTestHarness] closing 1 sockets
ERROR: Cannot schedule tasks on an EventLoop that has already shut down. This will be upgraded to a forced crash in future SwiftNIO versions.
ERROR: Cannot schedule tasks on an EventLoop that has already shut down. This will be upgraded to a forced crash in future SwiftNIO versions.
```

Without any connections it shutsdown immediately:

```
[ codes.vapor.application ] [ INFO ] run() (SocketServerTestHarness/SocketServerController.swift:135)
[ codes.vapor.application ] [ NOTICE ] Server starting on http://127.0.0.1:8080 (Vapor/HTTPServer.swift:296)
2022-09-20T08:09:34-0700 info 127.0.0.1:8080 : [SocketServerTestHarness] cancelling task...
[ codes.vapor.application ] [ DEBUG ] cancel handler called (SocketServerTestHarness/SocketServerController.swift:128)
[ codes.vapor.application ] [ DEBUG ] Application shutting down (Vapor/Application.swift:135)
[ codes.vapor.application ] [ TRACE ] Shutting down providers (Vapor/Application.swift:137)
[ codes.vapor.application ] [ TRACE ] Clearing Application storage (Vapor/Application.swift:141)
[ codes.vapor.application ] [ DEBUG ] Requesting HTTP server shutdown (Vapor/HTTPServer.swift:314)
[ codes.vapor.application ] [ DEBUG ] HTTP server shutting down (Vapor/HTTPServer.swift:320)
[ codes.vapor.application ] [ TRACE ] Shutting down EventLoopGroup (Vapor/Application.swift:149)
[ codes.vapor.application ] [ TRACE ] Application shutdown complete (Vapor/Application.swift:158)
2022-09-20T08:09:34-0700 info 127.0.0.1:8080 : [SocketServerTestHarness] closing 0 sockets
```

**Comment by Lukasa** [comment]
This looks to me like there's a bug somewhere in Vapor that is hit when they fail to stop the http server in a timely fashion. This can be seen by the error messages from NIO, complaining that Vapor is trying to use an event loop that it has shut down. I suspect that's an issue here.

**Comment by mkll** [comment]
Since the problem in question did not go away on its own over time, I recently came up with a naive and straightforward solution: in the aforementioned `SignalHandler`, I reinstall the signal handler every, say, 15 seconds. To be honest, I thought that some of the internal components of Vapor at some point "pulled the blanket over itself", and then its signal handler stopped working, as it did before. So I reinstall it again and again.

And even after this, in my configuration, the `SignalHandler` helps against freezes that occur immediately after the start, but, nevertheless, when the application runs for a long time, it again stops responding to `SIGTERM` from systemd.

But yesterday, while trying to figure out and collect information about how signals work in Linux, I came across a convenient [tool](https://codeberg.org/keks24/psig) that displays the status of signal handling by any process, and I found out an interesting thing that changes the picture somewhat.

Here's what it looks like for a normally running Vapor process:
```
PID: 49433
Name: smartload
UID: 111 111 111 111
GID: 1001 1001 1001 1001
Queued: 0/3691

Signals caught:
---------------
Signal 15: SIGTERM
Signal 11: SIGSEGV
Signal 8: SIGFPE
Signal 7: SIGBUS
Signal 4: SIGKILL
Signal 2: SIGINT

Signals blocked:
----------------
No signals found.

Signals pending (thread):
-------------------------
No signals found.

Signals pending (process):
--------------------------
No signals found.

Signals ignored:
----------------
Signal 13: SIGPIPE
```
After getting acquainted with `psig`, I almost immediately caught the situation when Vapor “froze” again and stopped responding to signals. So I had a whole minute to see what his signals were at that moment. That's what:
```
PID: 48111
Name: smartload
UID: 111 111 111 111
GID: 1001 1001 1001 1001
Queued: 0/3691

Signals caught:
---------------
Signal 11: SIGSEGV
Signal 8: SIGFPE
Signal 7: SIGBUS
Signal 4: SIGKILL

Signals blocked:
----------------
No signals found.

Signals pending (thread):
-------------------------
No signals found.

Signals pending (process):
--------------------------
No signals found.

Signals ignored:
----------------
Signal 15: SIGTERM
Signal 13: SIGPIPE
Signal 2: SIGINT
```
Take a look at the "ignored" section. Both SIGTERM and SIGINT are present there.

But why? As far as I understand, if a process does not respond to a signal, it could be for three different reasons:
1. He didn't even try. Didn't install the handler.
2. He tried, installed the handler, but then something went wrong. (This entire discussion is based on this assumption).
3. He explicitly included the signal in the list of ignored ones.

It is impossible for a signal to be accidentally ignored by a process. If a signal is in the _ignored_ list, it means that the process **has explicitly included it** there. And, as we see, some internal component of Vapor at some point clearly and deliberately includes SIGTERM in the ignore list.

So, this is not a bug, it's a feature! :)

**Comment by mkll** [comment]
Specifically in my project, the search showed the following occurrences of "makeSignalSource"
```
$ grep -rnw './' -e "makeSignalSource" --include \*.swift
./.build/checkouts/queues/Sources/Queues/QueuesCommand.swift:71:        func makeSignalSource(_ code: Int32) {
./.build/checkouts/queues/Sources/Queues/QueuesCommand.swift:72:            let source = DispatchSource.makeSignalSource(signal: code, queue: signalQueue)
./.build/checkouts/queues/Sources/Queues/QueuesCommand.swift:81:        makeSignalSource(SIGTERM)
./.build/checkouts/queues/Sources/Queues/QueuesCommand.swift:82:        makeSignalSource(SIGINT)
./.build/checkouts/swift-service-lifecycle/Sources/UnixSignals/UnixSignalsSequence.swift:85:                    dispatchSource: DispatchSource.makeSignalSource(signal: sig.rawValue, queue: UnixSignalsSequence.queue) as! DispatchSource,
./.build/checkouts/vapor/Sources/Vapor/Commands/ServeCommand.swift:87:        func makeSignalSource(_ code: Int32) {
./.build/checkouts/vapor/Sources/Vapor/Commands/ServeCommand.swift:88:            let source = DispatchSource.makeSignalSource(signal: code, queue: signalQueue)
./.build/checkouts/vapor/Sources/Vapor/Commands/ServeCommand.swift:97:        makeSignalSource(SIGTERM)
./.build/checkouts/vapor/Sources/Vapor/Commands/ServeCommand.swift:98:        makeSignalSource(SIGINT)
./Sources/App/Helpers/SignalHandler.swift:24:				let source = DispatchSource.makeSignalSource(signal: sig,  queue: .global())
```
The last line is my `SignalHandler` mentioned above. `Queue`? I don't use queues. So the following remain suspect: `UnixSignalsSequence` and `ServeCommand`.

**Comment by mkll** [comment]
But if you search for SIGTERM, the result is somewhat different:
```
$ grep -rnw './' -e "SIGTERM" --include \*.swift
admin@builder:~/downloaded/smartload$ grep -rnw './' -e "SIGTERM" --include \*.swift
./.build/checkouts/queues/Sources/Queues/QueuesCommand.swift:81:        makeSignalSource(SIGTERM)
./.build/checkouts/swift-service-lifecycle/Tests/UnixSignalsTests/UnixSignalTests.swift:121:        assert(.sigterm, rawValue: SIGTERM)
./.build/checkouts/swift-service-lifecycle/Tests/UnixSignalsTests/UnixSignalTests.swift:135:        assert(.sigterm, description: "SIGTERM")
./.build/checkouts/swift-service-lifecycle/Sources/UnixSignals/UnixSignal.swift:78:            return "SIGTERM"
./.build/checkouts/swift-service-lifecycle/Sources/UnixSignals/UnixSignal.swift:101:            return SIGTERM
./.build/checkouts/console-kit/Sources/ConsoleKitTerminal/Terminal/readpassphrase_linux.swift:47:        SIGTERM: .init(),   SIGTSTP: .init(),   SIGTTIN: .init(),   SIGTTOU: .init(),
./.build/checkouts/vapor/Sources/Vapor/Commands/ServeCommand.swift:5:/// Boots the application's server. Listens for `SIGINT` and `SIGTERM` for graceful shutdown.
./.build/checkouts/vapor/Sources/Vapor/Commands/ServeCommand.swift:97:        makeSignalSource(SIGTERM)
./Sources/App/Helpers/SignalHandler.swift:14:	let signals = [SIGTERM, SIGINT]
```

**Comment by gwynne** [comment]
If you actually look at `makeSignalSource()`, this is deliberate. The process-wide signal handler for the signal is set to `SIG_IGN` and responsibility for performing thread-specific signal masking is handed off to `DispatchSignalSource`.

**Comment by mkll** [comment]
@gwynne Actually, I don't have any in-depth knowledge about signal handling in Linux. But in this case, I want to make sure that I understand correctly: if we ignore SIGTERM at the process level, can it still be processed at the thread level?

And why doesn't Vapor, when just started, ignore SIGTERM if it should?

In general, I'm talking about the observed result. Namely, when Vapor ignores SIGTERM, it does not stop on a signal from the outside, which, in my opinion, looks logical. The process either ignores the signal or processes it; there cannot be halftones here. Or maybe?

Once again, I lack knowledge, but simple common sense tells me that if a signal is ignored at the process level, then it simply will not be delivered to any of the process threads, so there is no point in talking about thread-level processing, it does not will happen. Am I wrong here?

**Comment by gwynne** [comment]
Although hopefully the above-mentioned PR hopefully addresses the problem at long last, I'm going to leave this issue open for now, at least until someone confirms whether or not it actually helped 😅.

**Comment by mkll** [comment]
@gwynne Thanks! Previously, if I overridden the threadpool in the `configure()`, Vapor stopped responding to the `SIGTERM` immediately after launch. This is not happening now, and this is very significant. It looks like the problem has gone away.

But let's wait a few days and watch, just to be sure.

**Comment by mkll** [comment]
@gwynne Over the past time, there has not been a single case where Vapor refused to go on a shutdown. I'm sure the problem is completely resolved.

**Comment by 0xTim** [comment]
Wonderful! Will close this and we can reopen if we need to

**Comment by clarkezone** [comment]
This has also fixed it for me.  Thanks, great work!
